pkg/queue/cluster_queue_impl.go

```diff
@@ -55,39 +72,99 @@ func (c *ClusterQueueImpl) Cohort() string {
 func (c *ClusterQueueImpl) AddFromQueue(q *Queue) bool {
 	added := false
 	for _, info := range q.items {
-		if c.pushIfNotPresent(info) {
+		if c.heap.PushIfNotPresent(info) {
 			added = true
 		}
 	}
 	return added
 }
 
+func (c *ClusterQueueImpl) PushOrUpdate(wInfo *workload.Info) {
+	key := workload.Key(wInfo.Obj)
+	oldInfo := c.inadmissibleWorkloads[key]
+	if oldInfo != nil {
+		// update in place if the workload was inadmissible and didn't change
+		// to potentially become admissible.
+		if equality.Semantic.DeepEqual(oldInfo.Obj.Spec, wInfo.Obj.Spec) {
+			c.inadmissibleWorkloads[key] = wInfo
+			return
+		}
+		// otherwise move or update in place in the queue.
+		delete(c.inadmissibleWorkloads, key)
+	}
+	c.heap.PushOrUpdate(wInfo)
+}
+
+func (c *ClusterQueueImpl) Delete(w *kueue.Workload) {
+	delete(c.inadmissibleWorkloads, workload.Key(w))
+	c.heap.Delete(workload.Key(w))
+}
+
 func (c *ClusterQueueImpl) DeleteFromQueue(q *Queue) {
+	for _, w := range q.items {
+		key := workload.Key(w.Obj)
+		if wl := c.inadmissibleWorkloads[key]; wl != nil {
+			delete(c.inadmissibleWorkloads, key)
+		}
+	}
 	for _, w := range q.items {
 		c.Delete(w.Obj)
 	}
 }
 
-// pushIfNotPresent pushes the workload to ClusterQueue.
-// If the workload is already present, returns false. Otherwise returns true.
-func (c *ClusterQueueImpl) pushIfNotPresent(info *workload.Info) bool {
-	return c.heap.PushIfNotPresent(info)
-}
+// RequeueIfNotPresent inserts a workload that cannot be admitted into
+// ClusterQueue, unless it is already in the queue. If immediate is true,
+// the workload will be pushed back to heap directly. If not,
+// the workload will be put into the inadmissibleWorkloads.
+func (c *ClusterQueueImpl) RequeueIfNotPresent(wInfo *workload.Info, immediate bool) bool {
```

Comment:
I think it's better to implement this in `StrictFIFO` and `BestEffortFIFO` separately, I think `StrictFIFO` doesn't need the parameter `immediately` right? It can be more clearly, but yes some codes are redundant. Then we no longer need the `RequeueReason`.

I think namespaceSelector validation is always needed no matter `immediately` is true or not. For even immediately is true, workload with namespaceSelector not mach is still unschedulable, we shouldn't push it into the heap.